src/runtime/contrib/libxsmm/libxsmm_json_runtime.cc

```diff
@@ -0,0 +1,191 @@
+        // kernel.
+        if (has_bias_ || has_relu_) {
+          gemm_fusion_kernel_ = libxsmm_dispatch_brgemm_ext_v2(l_shape, l_flags, l_prefetch_flags,
+                                                               l_brconfig, l_argops, l_postops);
+        } else {
+          gemm_kernel_ = libxsmm_dispatch_brgemm_v2(l_shape, l_flags, l_prefetch_flags, l_brconfig);
+        }
+      }
+    }
+  }
+
+  void Run() override {
+    // Get input/output buffers.
+    auto data_eid = EntryID(input_nodes_[0], 0);
+    auto filter_eid = EntryID(input_nodes_[1], 0);
+    auto output_eid = EntryID(outputs_[0]);
+
+    void* data_handle = data_entry_[data_eid]->data;
+    void* filter_handle = data_entry_[filter_eid]->data;
+    void* output_handle = data_entry_[output_eid]->data;
+
+    // Transpose weight matrix since libxsmm only support GEMM rather than DENSE.
+    if (!transposed_filter_handle_) {
+      TVMDeviceAllocDataSpace(dev, K * N * sizeof(float), kAllocAlignment, type_hint,
+                              &transposed_filter_handle_);
+      for (int k = 0; k < K; ++k) {
+        for (int n = 0; n < N; ++n) {
+          static_cast<float*>(transposed_filter_handle_)[k * N + n] =
+              static_cast<float*>(filter_handle)[n * K + k];
+        }
+      }
+    }
+
+    if (has_bias_ || has_relu_) {
+      // Setup GEMM params.
+      libxsmm_gemm_ext_param gemm_param_ext;
+      gemm_param_ext.a.secondary = NULL;
+      gemm_param_ext.b.secondary = NULL;
+
+      gemm_param_ext.a.primary = transposed_filter_handle_;
+      gemm_param_ext.b.primary = data_handle;
+      gemm_param_ext.c.primary = output_handle;
+      if (has_bias_) {
+        auto bias_eid = EntryID(input_nodes_[2], 0);
+        void* bias_handle = data_entry_[bias_eid]->data;
+
+        gemm_param_ext.d.primary = bias_handle;
+      }
+      if (has_relu_) {
+        gemm_param_ext.c.secondary = relu_mask_.data();
+      }
+      gemm_param_ext.op.tertiary = &blocks_;
+
+      // Run GEMM fusion kernel.
+      gemm_fusion_kernel_(&gemm_param_ext);
+    } else {
+      // Setup GEMM params.
+      libxsmm_gemm_param gemm_param;
+      gemm_param.a.secondary = NULL;
+      gemm_param.b.secondary = NULL;
+
+      gemm_param.a.primary = transposed_filter_handle_;
+      gemm_param.b.primary = data_handle;
+      gemm_param.c.primary = output_handle;
+      gemm_param.op.tertiary = &blocks_;
+
+      // Run GEMM kernel.
+      gemm_kernel_(&gemm_param);
+    }
+  }
+
+  ~LibxsmmJSONRuntime() { TVMDeviceFreeDataSpace(dev, transposed_filter_handle_); }
+
+ private:
+  libxsmm_gemmfunction gemm_kernel_;
+  libxsmm_gemmfunction_ext gemm_fusion_kernel_;
+
+  // Transposed weight is saved to avoid redundant transpose in following steps.
+  // TODO(wenxizhu): check if current graph executor is in inference mode.
+  void* transposed_filter_handle_{nullptr};
+
+  DLDevice dev{kDLCPU, 0};
+  DLDataType type_hint{2, 32, 1};
+
+  int64_t M;
+  int64_t K;
+  int64_t N;
```

Comment:
Please follow the naming convention for private class members.

Reply:
Done.